src/Widgets/FirmwareUpdateWidget.vala

```diff
@@ -0,0 +1,176 @@
+            pixel_size = 32
+        };
+
+        var device_name_label = new Gtk.Label (device.name) {
+            halign = Gtk.Align.START,
+            hexpand = true
+        };
+        device_name_label.get_style_context ().add_class (Granite.STYLE_CLASS_H3_LABEL);
+
+        var version_label = new Gtk.Label (device.latest_release.version) {
+            wrap = true,
+            xalign = 0
+        };
+
+        var grid = new Gtk.Grid () {
+            column_spacing = 12,
+            margin = 6
+        };
+        grid.attach (icon, 0, 0, 1, 2);
+        grid.attach (device_name_label, 1, 0);
+        grid.attach (version_label, 1, 1);
+
+        switch (device.latest_release.flag) {
+            case ReleaseFlag.IS_UPGRADE:
+                if (device.latest_release.version == device.version) {
+                    add_up_to_date_label (grid);
+                    break;
+                }
+
+                var update_button = new Gtk.Button.with_label (_("Update")) {
+                    valign = Gtk.Align.CENTER
+                };
+                update_button.clicked.connect (() => {
+                    on_update_start ();
+
+                    update.begin (device, device.latest_release, (obj, res) => {
+                        update.end (res);
+                        on_update_end ();
+                    });
+                });
+                grid.attach (update_button, 2, 0, 1, 2);
+                break;
+            default:
+                add_up_to_date_label (grid);
+                break;
+        }
+
+        add (grid);
+    }
+
+    private void add_up_to_date_label (Gtk.Grid grid) {
+        var update_to_date_label = new Gtk.Label (_("Up to date")) {
+            valign = Gtk.Align.CENTER
+        };
+        grid.attach (update_to_date_label, 2, 0, 1, 2);
+    }
+
+    private async void update (Device device, Release release) {
+        var path = yield FwupdManager.get_instance ().download_file (release.uri);
+
+        var details = yield FwupdManager.get_instance ().get_details (device, path);
+
+        if (details.caption != null) {
+            show_details_dialog (details);
+        }
+
+        if ((yield FwupdManager.get_instance ().install (device, path)) == true) {
+            if (device.is (DeviceFlag.NEEDS_REBOOT)) {
+                show_reboot_dialog ();
+            } else if (device.is (DeviceFlag.NEEDS_SHUTDOWN)) {
+                show_shutdown_dialog ();
+            }
+        }
+    }
+
+    private void show_details_dialog (Details details) {
+        var message_dialog = new Granite.MessageDialog.with_image_from_icon_name (
+            _("Manual steps required"),
+            details.caption,
+            "application-x-firmware",
+            Gtk.ButtonsType.NONE
+        );
+        message_dialog.transient_for = (Gtk.Window) get_toplevel ();
+
+        var suggested_button = new Gtk.Button.with_label (_("OK"));
```

Comment:
We shouldn't use the label "OK" since it doesn't communicate a specific action. It's unclear from the button label what this will actually do.